framework/src/meshgenerators/SidesetsEncloseBlocks.C

```diff
@@ -0,0 +1,140 @@
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addParam<std::vector<SubdomainName>>(
+      "block", "The set of blocks that are checked if they are enclosed by boundary");
+  params.addParam<std::vector<BoundaryName>>("boundary",
+                                             "The name of the boundaries enclosing the ");
+  params.addParam<BoundaryName>("new_boundary", "The name of the boundary to create");
+  params.addClassDescription(
+      "MeshGenerator that checks if a set of blocks is enclosed by a set of sidesets."
+      "It can either add sides that are not covered by a sideset by a new sidesets or"
+      "error out.");
+
+  return params;
+}
+
+SidesetsEncloseBlocks::SidesetsEncloseBlocks(const InputParameters & parameters)
+  : MeshGenerator(parameters), _input(getMesh("input"))
+{
+}
+
+std::unique_ptr<MeshBase>
+SidesetsEncloseBlocks::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+
+  // error on finding a side that is not covered
+  bool error_out = !isParamValid("new_boundary");
+  boundary_id_type new_sideset_id;
+  if (!error_out)
+  {
+    BoundaryName new_boundary = getParam<BoundaryName>("new_boundary");
+    std::stringstream ss;
+    ss << new_boundary;
+    ss >> new_sideset_id;
+    if (ss.fail())
+      new_sideset_id = boundary_info.get_id_by_name(new_boundary);
+
+    if (new_sideset_id == BoundaryInfo::invalid_id)
+      paramError("new_boundary", "Not a valid boundary");
+  }
+
+  std::vector<subdomain_id_type> vec_block_ids =
+      MooseMeshUtils::getSubdomainIDs(*mesh, getParam<std::vector<SubdomainName>>("block"));
+  std::set<subdomain_id_type> blk_ids(vec_block_ids.begin(), vec_block_ids.end());
+
+  // get boundaries
+  // check if the provided sideset name is actually a sideset id
+  // if it can be converted to integer it's interpreted
+  // as a sideset id
+  auto boundary_name_vec = getParam<std::vector<BoundaryName>>("boundary");
+  std::vector<boundary_id_type> bnd_ids_vec;
+  bnd_ids_vec.reserve(boundary_name_vec.size());
+  std::set<boundary_id_type> bnd_ids;
+  for (auto & bnd_name : boundary_name_vec)
+  {
+    std::stringstream ss;
+    boundary_id_type sideset_id;
+    ss << bnd_name;
+    ss >> sideset_id;
+    if (ss.fail())
+      sideset_id = boundary_info.get_id_by_name(bnd_name);
+
+    if (sideset_id == BoundaryInfo::invalid_id)
+      paramError("boundary", "Not a valid boundary");
+    bnd_ids_vec.push_back(sideset_id);
+    bnd_ids.insert(sideset_id);
+  }
+
+  // loop over all elements in blocks and for each elem over each side
+  // and check the neighbors
+  for (auto & block_id : blk_ids)
+  {
+    for (const Elem * elem : as_range(mesh->active_local_subdomain_elements_begin(block_id),
+                                      mesh->active_local_subdomain_elements_end(block_id)))
+    {
```

Comment:
I think I would do something like the following, but I would have @roystgnr confirm
```suggestion
  for (const Elem * elem : as_range(mesh->local_level_elements_begin(0),
                                    mesh->local_level_elements_end(0)))
  {
    if (!blk_ids.count(elem->subdomain_id()))
      continue;
```
I would only bother looping over level 0 because in fact we error if you attempt to add sideset info for a non-level 0 element
```c++
void BoundaryInfo::add_side(const Elem * elem,
                            const unsigned short int side,
                            const boundary_id_type id)
{
  libmesh_assert(elem);

  // Only add BCs for level-0 elements.
  libmesh_assert_equal_to (elem->level(), 0);
```

Reply:
"You can't add sideset info for a non-level-0 element" is more of a missing feature than an API design.  We're hoping to fix that in the near future, once @fdkong or I have time to look at https://github.com/libMesh/libmesh/pull/3094 again.

"If you add boundary info to a parent element, that automatically associates all its descendant elements sharing that side with that boundary id", on the other hand, is an API design we'll be leaving unchanged.

Reply:
So say if I hit an element that was refined by AMR and I try to add one of its sides to a sideset, and I run in devel or dbg, then I would hit is assert (` libmesh_assert_equal_to (elem->level(), 0);`)? 

In that case, the code cannot safely work for AMR and potentially not for distributed mesh either. I know how to check for distributed mesh and error out. For AMR, can I check if the active element is the ultimate parent or something like that?

Reply:
Hmm... checking if an active element is level-0 (the ultimate parent) isn't a good AMR check (a uniformly refined mesh will have active elements at higher levels), but it might be a good check for your situation here, for now: when you're looping over active elements just error out if you're not at level 0.  Even on that hypothetical uniformly refined mesh, you won't currently be allowed to add boundary ids to child sides.

If we didn't have https://github.com/libMesh/libmesh/pull/3094 waiting in the wings I'd say it would be worth a little effort to try to handle refined meshes where child elements do all belong to their parents' subdomains, but instead the best thing for you to do might be to just handle the unrefined mesh case for now but tag yourself on that PR so we don't forget to notify you when handling the refined case too becomes an option.